File: backend/app/schemas/steps.py
"""Request + response shapes for /api/steps/*.

Read endpoints return a single, ready-to-display JSON object. Clients
should render the body directly; they should not aggregate, rank, or
filter on their side.

The write endpoint (POST /api/steps) accepts the minimum payload an
iOS Shortcut can build: a timestamp and a step count. `user_id` is
resolved from the Bearer token, never from the body.
"""

from datetime import date, datetime
from typing import Optional

from pydantic import BaseModel, Field


class LeaderboardEntry(BaseModel):
    rank: int
    username: str
    total: int


class DailyTotal(BaseModel):
    date: date
    total: int


class GlobalDailyResponse(BaseModel):
    date: date
    total_steps: int
    participating_users: int
    leaderboard: list[LeaderboardEntry]


class GlobalWeeklyResponse(BaseModel):
    week_start: date
    week_end: date  # inclusive; week_start + 6 days
    total_steps: int
    leaderboard: list[LeaderboardEntry]  # ranked by weekly total
    daily_breakdown: list[DailyTotal]  # 7 entries, one per day, in order


class SummaryLeader(BaseModel):
    username: str
    total: int


class BestDayEver(BaseModel):
    date: date
    total: int
    username: str


class GlobalSummaryResponse(BaseModel):
    total_users: int
    total_steps_all_time: int
    today_leader: Optional[SummaryLeader] = None
    this_week_leader: Optional[SummaryLeader] = None
    best_day_ever: Optional[BestDayEver] = None


class StepPost(BaseModel):
    timestamp: datetime
    total: int


class UserDailyResponse(BaseModel):
    username: str
    date: date
    total: int
    rank_today: Optional[int] = None  # null if user has no posts that day
    posts: list[StepPost]  # all snapshots, oldest first


class UserWeeklyResponse(BaseModel):
    username: str
    week_start: date
    week_end: date
    weekly_total: int
    rank_this_week: Optional[int] = None
    daily_breakdown: list[DailyTotal]


class UserBestDay(BaseModel):
    date: date
    total: int


class UserSummaryResponse(BaseModel):
    username: str
    join_date: datetime
    total_steps_all_time: int
    best_day: Optional[UserBestDay] = None
    rank_all_time: Optional[int] = None
    days_active: int


# ---------------------------------------------------------------------------
# Write
# ---------------------------------------------------------------------------


class CreateStepRequest(BaseModel):
    """Body shape for POST /api/steps. The iOS Shortcut sends one of
    these every time it syncs Apple Health step data."""

    timestamp: datetime
    total: int = Field(ge=0)
    # Upper bound is enforced as a sanity outlier in the service layer
    # (OUTLIER_CAP), not at the schema level — we want the row stored
    # so we can see what HealthKit sent, but excluded from aggregations.


class CreateStepResponse(BaseModel):
    """Row returned to the client after a successful POST."""

    id: int
    user_id: int
    timestamp: datetime
    total: int

File: frontend/src/api/steps.ts
import { apiFetch } from './client';

export interface LeaderboardEntry {
  rank: number;
  username: string;
  total: number;
}

export interface DailyTotal {
  date: string;
  total: number;
}

export interface GlobalDailyResponse {
  date: string;
  total_steps: number;
  participating_users: number;
  leaderboard: LeaderboardEntry[];
}

export interface GlobalWeeklyResponse {
  week_start: string;
  week_end: string;
  total_steps: number;
  leaderboard: LeaderboardEntry[];
  daily_breakdown: DailyTotal[];
}

export interface SummaryLeader {
  username: string;
  total: number;
}

export interface BestDayEver {
  date: string;
  total: number;
  username: string;
}

export interface GlobalSummaryResponse {
  total_users: number;
  total_steps_all_time: number;
  today_leader: SummaryLeader | null;
  this_week_leader: SummaryLeader | null;
  best_day_ever: BestDayEver | null;
}

export interface StepPost {
  timestamp: string;
  total: number;
}

export interface UserDailyResponse {
  username: string;
  date: string;
  total: number;
  rank_today: number | null;
  posts: StepPost[];
}

export interface UserWeeklyResponse {
  username: string;
  week_start: string;
  week_end: string;
  weekly_total: number;
  rank_this_week: number | null;
  daily_breakdown: DailyTotal[];
}

export interface UserMonthlyResponse {
  username: string;
  month_start: string;
  month_end: string;
  monthly_total: number;
  rank_this_month: number | null;
  daily_breakdown: DailyTotal[];
}

export interface UserBestDay {
  date: string;
  total: number;
}

export interface UserSummaryResponse {
  username: string;
  join_date: string;
  total_steps_all_time: number;
  best_day: UserBestDay | null;
  rank_all_time: number | null;
  days_active: number;
}

export function getGlobalDaily(date?: string): Promise<GlobalDailyResponse> {
  const qs = date ? `?date=${encodeURIComponent(date)}` : '';
  return apiFetch<GlobalDailyResponse>(`/steps/daily${qs}`);
}

export function getGlobalWeekly(
  weekStart?: string,
): Promise<GlobalWeeklyResponse> {
  const qs = weekStart ? `?week_start=${encodeURIComponent(weekStart)}` : '';
  return apiFetch<GlobalWeeklyResponse>(`/steps/weekly${qs}`);
}

export function getGlobalSummary(): Promise<GlobalSummaryResponse> {
  return apiFetch<GlobalSummaryResponse>('/steps/summary');
}

export function getUserDaily(
  username: string,
  date?: string,
): Promise<UserDailyResponse> {
  const qs = date ? `?date=${encodeURIComponent(date)}` : '';
  return apiFetch<UserDailyResponse>(
    `/steps/users/${encodeURIComponent(username)}/daily${qs}`,
  );
}

export function getUserWeekly(
  username: string,
  weekStart?: string,
): Promise<UserWeeklyResponse> {
  const qs = weekStart ? `?week_start=${encodeURIComponent(weekStart)}` : '';
  return apiFetch<UserWeeklyResponse>(
    `/steps/users/${encodeURIComponent(username)}/weekly${qs}`,
  );
}

export function getUserMonthly(
  username: string,
  month?: string,
): Promise<UserMonthlyResponse> {
  const qs = month ? `?month=${encodeURIComponent(month)}` : '';
  return apiFetch<UserMonthlyResponse>(
    `/steps/users/${encodeURIComponent(username)}/monthly${qs}`,
  );
}

export function getUserSummary(username: string): Promise<UserSummaryResponse> {
  return apiFetch<UserSummaryResponse>(
    `/steps/users/${encodeURIComponent(username)}/summary`,
  );
}
